src/sdk/namespace_fs.js

```diff
@@ -1390,35 +1398,89 @@ class NamespaceFS {
         }
     }
 
+    _get_part_data_path(params) {
+        return path.join(params.mpu_path, `parts-size-${params.size}`);
+    }
+
+    _get_part_md_path(params) {
+        return path.join(params.mpu_path, `part-${params.num}`);
+    }
+
+    // optimized version of upload_multipart - 
+    // 1. if size is pre known -
+    //    1.1. calc offset
+    //    1.2. upload data to by_size file in offset position
+    //    1.3. set on the part_md_file size, offset and etag xattr
+    // 2. else -
+    //    2.1. upload data to part_md_file
+    //    2.2. calc offset
+    //    2.3. copy the bytes to by_size file
+    //    2.4. set on the part_md_file size, offset and etag xattr
     async upload_multipart(params, object_sdk) {
-        // We can use 'wt' for open mode like we do for upload_object() when
-        // we figure out how to create multipart upload using temp files.
-        const open_mode = 'w';
+        let open_mode = 'r+';
+        const md_open_mode = 'w+';
+        const w_open_mode = 'w';
         const fs_context = this.prepare_fs_context(object_sdk);
         let target_file;
+        let part_md_file;
         try {
             await this._load_multipart(params, fs_context);
-            const upload_path = path.join(params.mpu_path, `part-${params.num}`);
-            // Will get populated in _upload_stream with the MD5 (if MD5 calculation is enabled)
+
+            const md_upload_path = this._get_part_md_path(params);
+            part_md_file = await this._open_file(fs_context, md_upload_path, md_open_mode);
+            let md_upload_params = {
+                fs_context, params, object_sdk, upload_path: md_upload_path, open_mode: md_open_mode,
+                target_file: part_md_file, file_path: md_upload_path
+            };
+
+            let upload_res;
+            const pre_known_size = params.size;
```

Comment:
change to boolean to allow size 0 parts (theoretic)
```suggestion
             pre_known_size = (size >= 0);
```